Newsletters; add 340 (2025-02-07) (#2157)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2025-02-07-newsletter.md

---
title: 'Bitcoin Optech Newsletter #340'
permalink: /en/newsletters/2025/02/07/
name: 2025-02-07-newsletter
slug: 2025-02-07-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME

## News

- **Channel force closure vulnerability in LDK:** Matt Morehouse
  [posted][morehouse forceclose] to Delving Bitcoin to announce a
  vulnerability affecting LDK that he [responsibly disclosed][topic

>   responsible disclosures] and which was fixed in LDK version 0.1.1.

Author: Keeping as-is.  I don't understand your suggestion; shouldn't there at least be a comma or emdash introducing the appositive phrase if the conjuction is removed?  I think it's better to just keep the conjunction.

  Similar to Morehouse's other recently disclosed vulnerability in LDK
  (see [Newsletter #339][news339 ldkvuln]), a loop in LDK's code
  terminated the first time it handled an unusual occurrence, preventing
  it from handling additional occurrences of the same problem.  In this
  case, it could result in LDK failing to settle pending [HTLCs][topic
  htlc] in open channels, ultimately leading honest counterparties to
  force close the channels so that they could settle the HTLCs onchain.

  This likely couldn't lead to direct theft, but it could result in the
  victim paying fees for the closed channels, paying fees to open new
  channels, and reduce the victim's ability to earn forwarding fees.

  Morehouse's excellent post goes into additional detail and suggests

>   how future bugs from the same underlying cause could be avoided.

Author: Changed to "might be avoided".  I wanted to use the hypothetical voice here but agree that "could" wasn't conveying it.

- **Zero-knowledge gossip for LN channel announcements:** Johan Halseth
  [posted][halseth zkgoss] to Delving Bitcoin with an extension to the
  proposed 1.75 [channel announcement][topic channel announcements]
  protocol that would allow other nodes to verify that a channel was
  backed by a funding transaction, preventing multiple cheap DoS
  attacks, but without revealing exactly which UTXO is the funding
  transaction to enhance privacy.  Halseth's extension builds on his
  previous research (see [Newsletter #321][news321 zkgoss]) that uses
  [utreexo][topic utreexo] and a zero-knowledge (ZK) proof system.  It
  would be applied to [MuSig2][topic musig]-based [simple taproot
  channels][topic simple taproot channels].

  The discussion focused on the tradeoffs between Halseth's idea,
  continuing to use non-private gossip, and alternative methods for
  generating the ZK proof.  Concerns included ensuring that
  all LN nodes can quickly verify proofs and the complexity of the proof
  and verification system since all LN nodes will need to include an
  implementation of it.

>   Discussion was ongoing at the time this summary was written.  <!-- FIXME:harding to review discussion again on Thursday -->

Author: Keeping as-is.  I think "time of this summary" is ambiguous, since the newsletter will be dated 2025-02-07 and that could be construed as the summary's time.  I think it's ok to use past tense for something we know will be read in the future.

- **Discovery of previous research for finding optimal cluster linearization:**
  Stefan Richter [posted][richter cluster] to Delving Bitcoin about a
  research paper from 1989 he found that has an proven algorithm which
  can be used to efficiently find the highest-feerate subset of a group

>   of transactions which remains topologically valid for inclusion in a

Author: Added extra words to make this even more verbose.  I think it's useful in a summary like this that will be read non-experts for us to use extra words to separate the different parts of the problem.  I think we can reasonably expect readers of Optech to understand what "highest feerate" means, but not all readers will understand what "topologically valid" means.  They might assume from "highest-feerate topologically-valid" that these two things automatically go together, like "blue jeans", but what we're really talking about here is ((highest feerate) and (topologically valid)), the analog of ((blue) and (jeans)).

  to immediately linearize a cluster, with a slower but more optimal
  algorithm able to run on spare CPU cycles.  However, if the 1989
  algorithm for the _maximum-ratio closure problem_, or another
  algorithm for that problem, is able to run fast enough, it could be
  used instead all of the time.  But, even if it was moderately slow, it
  could still be used as the algorithm that runs an spare CPU cycles.

  Pieter Wuille [replied][wuille cluster] with excitement and multiple
  follow-up questions.  He also [described][wuille sp cl] a new cluster
  linearization algorithm that the cluster mempool working group has
  been developing based on a discussion from Bitcoin Research Week with
  Dongning Guo and Aviv Zohar.  It converts the problem to one that can
  be addressed using [linear programming][] and appears to be fast, easy
  to implement, and produces an optimal linearization---if it
  terminates.  However, proof is needed that it will terminate (and in a
  reasonable amount of time).

  Although not directly related to Bitcoin, we found interesting
  Richter's [description][richter deepseek] of how he found the 1989

>   paper using the DeepSeek reasoning LLM.

Author: Keeping as-is.  Maybe my experience is atypical, but it seems to me that in everyday American usage, a sentence that ends in "interesting" is often meant sarcastically, e.g. "Larry's suggestion is... interesting".  I think our genuine interest is less easily misinterpreted if explicitly say that's how we felt.

  topics in computer vision."

- **Erlay update:** Sergi Delgado made several posts to Delving Bitcoin
  about his work over the past year implementing [Erlay][topic erlay]
  for Bitcoin Core.  He starts with an [overview][delgado erlay] of how
  existing transaction relay (called _fanout_) works and how erlay
  proposes to change that.  He notes that some amount of fanout is
  expected to remain even in a network where every node supports Erlay,
  as "fanout is more efficient and considerably faster than set
  reconciliation, provide the receiving node does not know about the
  transaction being announced."

  Using a mix of fanout and Erlay reconciliation requires choosing when
  to use each method and which peers to use them with, so his research
  has focused on making the optimal choices:

  - [Filtering based on transaction knowledge][sd1] examines whether a
    node should include a peer in its plans to fanout a transaction even
    if it knows that peer already has the transaction.  For example, our
    node has ten peers; three of those peers have already announced a
    transaction to us.  If we want randomly select three peers to

>     further fanout that transaction, should we select from all ten peers

Author: Note: changed the earlier "select" in this sentence to "choose", as that sounded better to me.  Thanks!

    should choose which peers will receive a fanout transaction (with the
    remainder using Erlay reconciliation).  Two options are considered:
    when the node processes the transaction and queues it for relay, and
    when it's time to be relayed (nodes wait a small random amount of
    time to relay transactions to make it harder to probe the network
    topology and guess which node originated a transaction, which would
    be bad for privacy).  Again, simulation results show that "there are
    no meaningful differences", although the "results may vary [...] in
    networks with partial [Erlay] support".

  - [How many peers should receive a fanout?][sd3] examines the fanout
    rate.  A higher rate accelerates transaction propagation but reduces
    bandwidth savings.  Besides testing the fanout rate, Delgado also
    tested increasing the number of outbound peers, as that's one of the
    goals of adopting Erlay.  The simulation shows the current Erlay
    approach reduced bandwidth by approximately 35% using current outbound
    peer limits (8 peers), and 45% using 12 outbound peers.  However,
    the transaction latency is increased by about 240%.  Many other

>     tradeoffs are graphed in the posted.  In addition to the results

Reviewer: ```suggestion
    tradeoffs are graphed in the post.  In addition to the results
```

    might be able to make better tradeoffs.

- **Tradeoffs in LN ephemeral anchor scripts:** Bastien Teinturier
  [posted][teinturier ephanc] to Delving Bitcoin to ask for opinions
  about what [ephemeral anchor][topic ephemeral anchors] script should
  be used as one of the outputs to [TRUC][topic v3 transaction
  relay]-based LN commitment transactions as a replacement for existing
  [anchor outputs][topic anchor outputs].  The script used determines
  who can [CPFP fee bump][topic cpfp] the commitment transaction (and
  under what conditions they can bump).  He presented four options:

  - _Use a pay-to-anchor (P2A) script:_ this has a minimal onchain size
    but means that all [trimmed HTLC][topic trimmed htlc] value will
    probably go to miners (like it presently does).

  - _Use a single-participant keyed anchor:_ this may allow some excess
    trimmed HTLC value to be claimed by a participant who voluntarily
    accepts waiting a few dozens blocks before being able to spend the
    money they close out of a channel.  Anyone who wants to force close
    a channel would need to wait that time anyway.  However, neither

>     party to the channel can delegate paying the fee to a third-party

Author: Keeping as-is.  I think "fee payment" is ambiguous about whether it's the fee or payment of the fee that's being delegated.

    without allowing that party to steal all of the channel funds.
    If both you and your counterparty compete to claim the excess value,
    it's likely that it will all go to miners anyway.

  - _Use a shared key anchor:_ this also allows recovery of excess
    trimmed HTLC value and allows delegation, but anyone who receives
    the delegation can compete with you and your counterparty to claim
    the excess value.  Again, in the case of competition, all excess
    value is likely to go to miners.

  - _Use a dual-keyed anchor:_ allows each participant to claim excess
    trimmed HTLC value without having to wait additional blocks before
    being able to spend.  However, it does not allow delegation.  The
    two parties to a channel can still compete with each other.

  In replies to the post, Gregory Sanders [noted][sanders ephanc] that
  the different schemes could be used at different times.  For example,
  P2A could be used when there were no trimmed HTLCs and one of the
  keyed anchors could be used other times.  He also suggested that
  trimmed value above the [dust threshold][topic uneconomical outputs]

>   could instead be added to the LN commitment transaction.

Author: Yeah, that change wouldn't be quite right, but I can see that the "instead" in the sentence was ambigous.  Made a different change that hopefully clarifies.

>   Additionally, we warned that it could create "new weirdness [a]

Reviewer: ```suggestion
  Additionally, he warned that it could create "new weirdness [a]
```

  counterparty may be tempted to ramp up the trimmed amount and take it
  themselves."  David Harding extended on this theme in a [later
  post][harding ephanc].

  Antoine Riard [warned][riard ephanc] against using P2A due to the risk
  of encouraging miner [transaction pinning][topic transaction pinning]
  (see [Newsletter #339][news339 pincycle]).

  Discussion was ongoing at the time of writing.

- **Emulating OP_RAND:** Oleksandr Kurbatov [posted][kurbatov rand] to
  Delving Bitcoin about an interactive protocol that allows two parties
  to make a contract that will pay out in a way neither of them can

>   predict, in other words: "randomly".  [Previous work][dryja pp] on

Reviewer 2: ```suggestion
  predict, in other words, randomly.  [Previous work][dryja pp] on
```
(maybe I'm missing a reason "randomly" is in quotes)

Author: Removed quotes and added words to explain that this isn't really random but it's equivalent for this case.  It's a bit verbose but it'll cut down on the number of complaints that I receive from pedants.  :-)

  _probabilistic payments_ in Bitcoin has used advanced scripts, but

>   Kurbatov's approach entirely uses specially constructed public keys

Reviewer 2: ```suggestion
  Kurbatov's approach uses specially-constructed public keys
```
(or "uses only specially-constructed")

Author: Removed "entirely" but didn't add the dash.  My understanding is that words ending in -ly don't need a dash following them when used as an adjectival phrase.

>   that allow the winner to spend the contracted funds.  This is more

Reviewer 2: ```suggestion
  allowing the winner to spend the contracted funds.  This is more
```
or, "that allows the winner..."

  private and may allow greater flexibility.

  Optech wasn't able to analyze the protocol, but we didn't spot any
  obvious problems.   We hope to see additional discussion of the
  idea---probabilistic payments have multiple applications, including
  allowing users to send amounts onchain that would otherwise be
  [uneconomical][topic uneconomical outputs], such as for [trimmed
  HTLCs][topic trimmed htlc].

- **Discussion about lowering the minimum transaction relay feerate:**
  Greg Tonoski [posted][tonoski minrelay] to the Bitcoin-Dev mailing
  list about lowering the [default minimum transaction relay
  feerate][topic default minimum transaction relay feerates], a topic
  that has been discussed repeatedly (and summarized by Optech) since
  2018---most recently in 2022 (see [Newsletter #212][news212 relay]).
  Of note, a recently disclosed vulnerability (see [Newsletter
  #324][news324 largeinv]) did reveal a potential problem that could
  have affected users and miners who lowered the setting in the past.

>   Optech will provide updates if new significant discussion occurs.

Reviewer 2: ```suggestion
  Optech will provide updates if significant further discussion occurs.
```
or maybe "future discussion"?

    vulnerability to produce blocks faster than once every 10 minutes.

    Poinsot now accepts using a 7,200 second (2 hour) grace period, as
    originally proposed by Sjors Provoost as being much less likely to
    lead to a miner accidentally producing an invalid block, although it
    does allow a very patient attacker controlling more than 50% of
    network hashrate can use the timewarp attack to lower difficulty
    over a period of months even if actual hashrate remains constant or
    increases.  This would be a publicly visible attack and the network
    would have months to react.  In his post, Poinsot summarizes
    previous arguments (see [Newsletter #335][news335 cc] for our much
    less detailed summary) and concludes that, "despite the fairly weak
    arguments in favor of increasing the grace period, the cost of doing
    so [does] not prohibit erring on the safe side."

    On a thread dedicated to discussing extending the grace period,
    developers Zazy and Pieter Wuille [discussed][wuille erlang] how the
    600 second grace period, which would seem to allow a small amount of
    time warping, was actually sufficient to ensure constant difficulty

>     under constant hashrate.  Specifically, they looked at the impact of

Reviewer 3: Under constant hashrate **and** under attack. Pieter's point is that the 600 grace period makes it so the worst case under an attack only increases the block rate to what it was originally advertized to be (1 block per 10 minutes on average).

## Notable code and documentation changes

_Notable recent changes in [Bitcoin Core][bitcoin core repo], [Core
Lightning][core lightning repo], [Eclair][eclair repo], [LDK][ldk repo],
[LND][lnd repo], [libsecp256k1][libsecp256k1 repo], [Hardware Wallet
Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
Server][btcpay server repo], [BDK][bdk repo], [Bitcoin Improvement
Proposals (BIPs)][bips repo], [Lightning BOLTs][bolts repo],
[Lightning BLIPs][blips repo], [Bitcoin Inquisition][bitcoin inquisition
repo], and [BINANAs][binana repo]._

- [Bitcoin Core #21590][] implements a safegcd-based [modular
  inversion][modularinversion] algorithm for MuHash3072 (see Newsletter
  [#131][news131 muhash]), based on libsecp256k1’s implementation while adding
  support for both 32-bit and 64-bit architectures and specializing for the
  specific modulus. Benchmark results show an approximate 100× performance
  improvement on x86_64, reducing MuHash’s computation from 5.8 ms to 57 μs,
  paving the way for more efficient state validation. The implementation is
  originally based on [Bernstein and Yang’s constant-time variant of Euclid’s

>   algorithm][bernsteinandyang].

Author: @Gustavojfe FYI: I'm going to delete this final sentence mentioning Bernstein and Yang.  Our soft policy is not to mention contributor names in this section due to merges being the result of the contributions of reviewers as much (if not more) than PR authors.  Although B&Y aren't Bitcoin contributors and this is more of an academic citation, I think a full citation here would credit the significant proving and performance optimizations made to SafeGCD by Wuille and Maxwell, who are Bitcoin contributors.  Instead of the current citation or mentioning W&M's work, I'll add a link to a previous newsletter where we previously discussed all of this (and more).

    topology and guess which node originated a transaction, which would
    be bad for privacy).  Again, simulation results show that "there are
    no meaningful differences", although the "results may vary [...] in
    networks with partial [Erlay] support".

  - [How many peers should receive a fanout][sd3] examines the fanout
    rate.  A higher rate accelerates transaction propagation but reduces
    bandwidth savings.  Besides testing the fanout rate, Delgado also
    tested increasing the number of outbound peers, as that's one of the
    goals of adopting Erlay.  The simulation shows the current Erlay
    approach reduced bandwidth by approximately 35% using current outbound
    peer limits (8 peers), and 45% using 12 outbound peers.  However,
    the transaction latency is increased by about 240%.  Many other
    tradeoffs are graphed in the post.  In addition to the results
    being helpful for choosing current parameters, Delgado notes that
    they'll be useful for evaluating alternative fanout algorithms that
    might be able to make better tradeoffs.

  - [Defining the fanout rate based on how a transaction was received][sd4]

>     examines whether the fanout range should be adjusted depending on

Reviewer 4: ```suggestion
    examines whether the fanout rate should be adjusted depending on
```

    two parties to a channel can still compete with each other.

  In replies to the post, Gregory Sanders [noted][sanders ephanc] that
  the different schemes could be used at different times.  For example,
  P2A could be used when there were no trimmed HTLCs, and one of the
  keyed anchors could be used other times.  If the
  trimmed value was above the [dust threshold][topic uneconomical
  outputs], that value could be added to the LN commitment
  transaction instead of an anchor output.
  Additionally, he warned that it could create "new weirdness [a]
  counterparty may be tempted to ramp up the trimmed amount and take it
  themselves."  David Harding extended on this theme in a [later
  post][harding ephanc].

  Antoine Riard [warned][riard ephanc] against using P2A due to the risk
  of encouraging miner [transaction pinning][topic transaction pinning]
  (see [Newsletter #339][news339 pincycle]).

  Discussion was ongoing at the time of writing.

- **Emulating OP_RAND:** Oleksandr Kurbatov [posted][kurbatov rand] to
  Delving Bitcoin about an interactive protocol that allows two parties
  to make a contract that will pay out in a way that neither can

>   predict, which is functionally equivilant to randomly.  [Previous work][dryja pp] on

Reviewer 5: ```suggestion
  predict, which is functionally equivalent to randomly.  [Previous work][dryja pp] on
```

    negative impacts of consensus solutions to the [duplicate
    transactions][topic duplicate transactions] problem, Poinsot
    selected a specific solution to include in the cleanup proposal:
    requiring the height of the previous block be included in each
    coinbase transaction's time lock field.  This proposal has two
    advantages: it allows [extracting][corallo duplocktime] the
    committed block height from a block without having to parse Script
    and it allows [creating][harding duplocktime] a compact SHA256-based
    proof of block height (about 700 bytes in the worst case, much less
    than the worst-case 1 MB proof that would be required now without an
    advanced proof system).

    This change will not affect regular users but it will eventually
    require miners to update the software they use to generate coinbase
    transactions.  If any miners have concerns about this proposal,
    please contact Poinsot or another protocol developer.

  Poinsot also [posted][ap4] a high-level update on his work and the
  proposal's current state to the Bitcoin-Dev mailing list.

- **Request for a covenant design supporting Braidpool:** Bob McElrath

>   [posted][mcelrath braidcov] to Delving Bitcoin requestesting developers

Reviewer 4: ```suggestion
  [posted][mcelrath braidcov] to Delving Bitcoin requesting developers
```

  default (improving privacy by allowing BDK transactions to blend in
  with other wallets that must use v2 transactions due to their support
  for relative locktimes, see [Newsletter #337][news337 bdk]).  It also
  adds support for [compact block filters][topic compact block filters]
  (see [Newsletter #339][news339 bdk-cpf]), in addition to "various bug
  fixes and improvements".

- [LND v0.18.5-beta.rc1][] is a release candidate for a minor version of
  this popular LN node implementation.

## Notable code and documentation changes

_Notable recent changes in [Bitcoin Core][bitcoin core repo], [Core
Lightning][core lightning repo], [Eclair][eclair repo], [LDK][ldk repo],
[LND][lnd repo], [libsecp256k1][libsecp256k1 repo], [Hardware Wallet
Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
Server][btcpay server repo], [BDK][bdk repo], [Bitcoin Improvement
Proposals (BIPs)][bips repo], [Lightning BOLTs][bolts repo],
[Lightning BLIPs][blips repo], [Bitcoin Inquisition][bitcoin inquisition
repo], and [BINANAs][binana repo]._

- [Bitcoin Core #21590][] implements a safegcd-based [modular
  inversion][modularinversion] algorithm for MuHash3072 (see Newsletters

>   [#131][news131 muhash] and [#136][news136 gcd]), based on libsecp256k1’s implementation while adding
>   support for both 32-bit and 64-bit architectures and specializing for the
>   specific modulus. Benchmark results show an approximate 100× performance

Reviewer 4: I would suggest dropping the two mentions of differences between the original libsecp256k1 implementation and the new one. The first is imprecise as it is worded and the second is a bit redundant:

  - libsecp256k1 had support for both archs, but it was split into two files, one for each arch; the change is that this code is now generic for both archs.
  - the specialization is for MuHash3072, but that is already mentioned.

```suggestion
  [#131][news131 muhash] and [#136][news136 gcd]), translated into C++ from libsecp256k1’s code. Benchmark results show an approximate 100× performance
```
